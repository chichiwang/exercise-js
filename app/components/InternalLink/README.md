# InternalLink
A simple higher-order component that wraps the [React Router Link](https://reactrouter.com/en/main/components/link) component, ensuring that a [state](https://reactrouter.com/en/main/components/link#state) object property `resetScroll` is set to `true` on the Link component.

This is a convenience component to ensure client-side navigation will reset the browser window scroll on page navigation. This issue is discussed in greater detail in the hook [app/hooks/useResetScroll](app/hooks/README.md#useResetScroll).
