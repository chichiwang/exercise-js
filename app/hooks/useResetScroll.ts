import { useLayoutEffect } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

export const ResetScrollKey = 'resetScroll';

function useResetScroll() {
  const location = useLocation();
  const navigate = useNavigate();

  useLayoutEffect(function resetScrollPosition() {
    if (location.state?.[ResetScrollKey]) {
      const IdFromHash = location.hash.slice(1);
      const $hashEl = document.getElementById(IdFromHash);

      if ($hashEl) {
        $hashEl.scrollIntoView(true);
      } else {
        document.documentElement.scrollTo({
          top: 0,
          left: 0,
          behavior: 'instant',
        });
      }

      navigate(`${location.pathname}${location.hash}`, {
        replace: true,
      });
    }
  }, [location.pathname, location.hash]);
}

export default useResetScroll;
