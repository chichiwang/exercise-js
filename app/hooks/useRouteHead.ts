import { useLocation, useMatches } from 'react-router-dom';
import type { PageMatch, RouteHead } from 'app/types';

function useRouteHead(): RouteHead {
  const location = useLocation();
  const routeMatches = useMatches();

  const locationMatch = routeMatches.find(
    function matchLocationWithHead(match): match is PageMatch {
      const pathMatches = match.pathname === location.pathname;
      const containsHandle = typeof match.handle === 'object';
      const containsHead = containsHandle
        && Object.prototype.hasOwnProperty.call(match.handle, 'head');

      return pathMatches && containsHead;
    },
  );

  const pathChainMatches = routeMatches.filter(
    function matchesWithHead(match): match is PageMatch {
      const containsHandle = typeof match.handle === 'object';

      return containsHandle && Object.prototype.hasOwnProperty.call(match.handle, 'head');
    },
  );

  const pathChainMatch = pathChainMatches.length
    ? pathChainMatches[pathChainMatches.length - 1]
    : undefined;

  return locationMatch?.handle?.head || pathChainMatch?.handle?.head || {};
}

export default useRouteHead;
