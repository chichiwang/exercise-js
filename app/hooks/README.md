# Hooks
This directory houses [custom React hooks](https://react.dev/learn/reusing-logic-with-custom-hooks) created for the boilerplate.

## useResetScroll

[useResetScroll](./useResetScroll.ts) is a hook created specifically for Layout components that utilize the [React Router Outlet](https://reactrouter.com/en/main/components/outlet). [Links](https://reactrouter.com/en/main/components/link) that navigate between pages under the same layout do not reset the browser window's scroll position upon navigation. To remedy this, this hook will be used to explicitly trigger a scroll reset back to the top-left of the browser window upon route change.

This scroll reset only triggers when a location change carries with it a specific flag in a location [state](https://reactrouter.com/en/main/components/link#state) object:

```javascript
import { Link } from 'react-router-dom';

<Link to="/internal-path" state={{ resetScroll: true }}>Reset scroll on navigation</Link>
```

This state object flag, `resetScroll`, can be applied automatically by using the reusable component [app/components/InternalLink](../components/InternalLink):

```javascript
import InternalLink from 'app/components/InternalLink';

<InternalLink to="/internal-path">Reset scroll on Navigation</InternalLink>
```

After resetting the scroll, this hook will replace the current [location state](https://reactrouter.com/en/main/start/concepts#definitions) with the same path removing the `resetScroll` state. The result of this is:
* Upon clicking a link with the `resetScroll` state will navigate the page and reset the page scroll.
* Using browser navigation (back/forward buttons) to return to a page that had its scroll reset will not reset the scroll, preserving the user's scroll position on the page.

### Hash Links
`useResetScroll` will behave differently when the `resetScroll` flag is present in the location state if the location contains a hash: instead of scrolling to the top-left of the window, it will instead scroll the element whose `id` attribute matches the hash into view.

```javascript
import InternalLink from 'app/components/InternalLink';

<InternalLink to="/internal-path#page-section">
  Scroll to element with id "page-section" on navigation
</InternalLink>
```

If the target page does not contain an element whose `id` attribute matches the location hash, this hook will scroll the browser to the top-left of the window instead.
