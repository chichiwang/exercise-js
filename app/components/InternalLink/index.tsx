import React from 'react';
import { Link } from 'react-router-dom';
import type { LinkProps } from 'react-router-dom';

import { ResetScrollKey } from 'app/hooks/useResetScroll';

interface InternalLinkProps extends Omit<LinkProps, 'state'> {
  state?: { [key: string]: any };
}

function InternalLink(props: InternalLinkProps) {
  const resetScrollState = {
    [ResetScrollKey]: true,
  };

  const { state } = props;

  const propsWithState = {
    ...props,
    ...(typeof state === 'undefined' ? {
      state: resetScrollState,
    } : {
      state: {
        ...state,
        ...resetScrollState,
      },
    }),
  };

  return (
    <Link {...propsWithState} />
  );
}

export default InternalLink;
