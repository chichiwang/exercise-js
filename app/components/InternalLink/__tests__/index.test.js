import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { ResetScrollKey } from 'app/hooks/useResetScroll';
import InternalLink from '../index';

function RenderProp({ k, v }) {
  return (
    <p data-testid={`${k}`}>{v}</p>
  );
}

function MockedLink(props) {
  return (
    <div>
      { Object.entries(props).map(function renderEachProp([key, value]) {
        const val = JSON.stringify(value);
        return <RenderProp key={key} k={key} v={val} />;
      }) }
    </div>
  );
}

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Link: MockedLink,
}));

const expectedStateContent = {
  [ResetScrollKey]: true,
};

describe('InternalLink component', () => {
  test('matches the snapshot', () => {
    const tree = renderer.create(<InternalLink to="/test" />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('passes all properties through to React Router Link', () => {
    const props = {
      to: '/test-props',
      children: 'nested-elements',
      foo: 'bar',
      hello: 'compositional component!',
    };

    render(<InternalLink {...props} />);

    for (const [key, value] of Object.entries(props)) {
      const $el = screen.queryByTestId(key);

      expect($el).toBeInTheDocument();
      expect($el).toHaveTextContent(JSON.stringify(value));
    }
  });

  test('adds a state object to trigger scroll reset', () => {
    const targetPath = '/test-state-applied';

    render(<InternalLink to={targetPath} />);

    expect(screen.queryByTestId('to')).toHaveTextContent(targetPath);
    expect(screen.queryByTestId('state')).toHaveTextContent(JSON.stringify(expectedStateContent));
  });

  test('extends provided state object with trigger for scroll reset', () => {
    const targetPath = '/test-state-extended';
    const initialStateObject = {
      some: 'state',
    };

    render(<InternalLink to={targetPath} state={initialStateObject} />);

    expect(screen.queryByTestId('to')).toHaveTextContent(targetPath);
    expect(screen.queryByTestId('state')).toHaveTextContent(JSON.stringify({
      ...initialStateObject,
      ...expectedStateContent,
    }));
  });
});
