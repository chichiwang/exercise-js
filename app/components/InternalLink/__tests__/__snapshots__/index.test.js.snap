// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalLink component matches the snapshot 1`] = `
<div>
  <p
    data-testid="to"
  >
    "/test"
  </p>
  <p
    data-testid="state"
  >
    {"resetScroll":true}
  </p>
</div>
`;
